src/mscorlib/shared/System/Globalization/TextInfo.cs

```diff
@@ -258,6 +258,43 @@ private unsafe string ToLowerAsciiInvariant(string s)
             }
         }
 
+        internal void ToLowerAsciiInvariant(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(destination.Length >= source.Length);
+
+            if (source.Length == 0)
+            {
+                return;
+            }
+
+            int i = 0;
+            while (i < source.Length)
+            {
+                if ((uint)(source[i] - 'A') <= ('Z' - 'A'))
+                {
+                    break;
+                }
+                i++;
+            }
+            
+            if (i >= source.Length)
+            {
+                source.CopyTo(destination);
+                return;
+            }
+
+            source.Slice(0, i).CopyTo(destination);
+
+            destination[i] = (char)(source[i] | 0x20);
+            i++;
+
+            while (i < source.Length)
```

Comment:
The string-based implementation of this has two separate loops like this because it wants to optimize the path where nothing needs to change and thus nothing needs to allocate.  But here with span, we always need to look at every character and copy every character, potentially modified, to the output destination.  Can we just consolidate it all to a single loop then? e.g.
```C#
internal void ToLowerAsciiInvariant(ReadOnlySpan<char> source, Span<char> destination)
{
    Debug.Assert(destination.Length >= source.Length);
    for (int i = 0; i < source.Length; i++)
    {
        destination[i] = ToLowerAsciiInvariant(source[i]);
    }
}
```
or is there some reason that's not as good?

Reply:
What about the optimization which avoids having to do the `or by 0x20` operation where not necessary (i.e when `((uint)(source[i] - 'A') > ('Z' - 'A'))`?

I suspect the extra CopyTo probably hurts performance more, so I will stick to the simpler implementation. I will measure the perf and compare.

Reply:
> What about the optimization

That's in the char method as well:
https://github.com/dotnet/coreclr/pull/16379/files/59903ebba9bd23fa952f834e41c87bfe3b726b85#diff-64100f763acd6161cac733b7534a9301R381
We'd just want to make sure that method gets inlined.